File: python/2015/day01.py
from common.constants import INPUT_DIR


def part2(data: list[str]) -> int:
    floor_num = 0
    for i, char in enumerate(data):
        if floor_num < 0:
            return i

        if char == '(':
            floor_num += 1
        else:
            floor_num -= 1

    return 0


if __name__ == '__main__':
    with open(INPUT_DIR + '2015/day01.txt', 'r') as file:
        data = list(file.read().strip())

    print(f"Part 1: {len(data) - 2 * data.count(')')}")
    print(f"Part 2: {part2(data)}")

File: c/2015/day01.c
#include "../common/utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int part1(const char *input);
int part2(const char *input);

int main() {
    char *input = read_file("../inputs/2015/day01.txt");
    if (!input) return 1;

    printf("Part 1: %d\n", part1(input));
    printf("Part 2: %d", part2(input));

    free(input);
    return 0;
}

int part1(const char *input) {
    int count = 0;
    const char *tmp = input;
    while ((tmp = strstr(tmp, "("))) {
        count++;
        tmp++;
    }

    return 2 * count - strlen(input);
}

int part2(const char *input) {
    int floor = 0;
    for (int i = 0; i < strlen(input); i++) {
        if (floor < 0) return i;

        if (input[i] == '(') {
            floor++;
        } else {
            floor--;
        }
    }

    return 0;
}
